Generate mcsema2 ABI Library definitions

artemdinaburg opened the pull request:
Include a script to generate ABI library files for mcsema2 from a header file.

* Add support for `__restrict__` to cparser
* Script to generate C++ that can be built to bitcode and used with mcsema2's `--abi-library`
* Some rudimentary test data

pgoodman commented on make_abi_library.py:
```diff
@@ -0,0 +1,211 @@
+STATS = {}
+
+
+def has_function(args):
```
I think what you should add a `seen` set argument to use alongside the global `FUNCTION_CACHE`. The reason why is that the initial assumption of `FUNCTION_CACHE[arg] = False` is there to stop recursion, but you don't want to prematurely determine that something else doesn't have a function pointer. For example:

```c
struct X;
typedef struct X Y;
struct X {
  Y *next;
  void (*foo)(void);
};
void bar(struct X *z);
void baz(Y *z);
````

I think in the above case, `baz` will be treated as not taking a function pointer, when really it does. So what you want is something like:

```python
def has_function(*args, seen=None):
  global FUNCTION_CACHE
  if seen is None:
    seen = set()
  for arg in args:
    # check FUNCTION_CACHE here, return whatever is in the FUNCTION_CACHE if present
    # check seen cache here, return False if seen
    # add to seen cache, but NOT function cache
    # if statements as below, passing `seen` down into calls of `has_function`, updating FUNCTION_CACHE with the result to be returned
```

By using `*args` instead of `args`, you can do `has_function(arg.ctype, seen=seen)`, and in the tuple case, do `has_function(*arg, seen=seen)`

pgoodman commented on make_abi_library.py:
```diff
@@ -0,0 +1,211 @@
+  # None of these args are a function
+  return False
+
+def has_function_pointer(args):
```
What is the purpose of this function? Why not just use `has_function`?

artemdinaburg commented:
Some comments:
1) After reading it carefully, `has_function_pointer` was indeed a no-op; removed
2) Migrated to `*args` syntax
3) Added the `seen()` set. It still triggered the false negative in your example, the easiest way to fix it was to remove caching of False values. The problem is that once you hit a value in `seen` and return False, all elements on the path to reach the value (but not the value itself) will be cached as False. Eventually you would reach the Value via another path, and mark it as true, but the intermediate elements stay at False and cause false negatives. It was easier to just remove the caching of False values, as the script is fast enough, even for libc.

pgoodman commented on make_abi_library.py:
```diff
@@ -15,34 +12,41 @@
+    # check if we have a cached determination
     cached = FUNCTION_CACHE.get(arg, None)
     if cached is not None:
       if cached == True:
```
I think you want `return cached` when the `cached is not None` is `True`.

artemdinaburg replied:
So cached would only ever be `True` for now, but your version is easier to understand and less code; will change.

pgoodman commented on make_abi_library.py:
```diff
@@ -15,34 +12,41 @@
 
 
-def has_function(args):
+def has_function(seen=None, *args):
```
Maybe rename this to:

```python
def _has_function(seen, *args):
  ...

def has_function(*args):
  global FUNCTION_CACHE
  has_func = False
  for ctype in args:
    if ctype not in FUNCTION_CACHE:
      FUNCTION_CACHE[ctype] = _has_function(set(), ctype)
    has_func = has_func or FUNCTION_CACHE[ctype]
  return has_func
```

pgoodman replied:
With this setup, you would no longer need to check if `seen is None`, and inside `_has_function`, you'd also call `_has_function` and you wouldn't assign to `FUNCTION_CACHE`.

artemdinaburg replied:
updated

pgoodman commented on make_abi_library.py:
```diff
@@ -0,0 +1,153 @@
+  for arg in args:
+    if arg not in FUNCTION_CACHE:
+      FUNCTION_CACHE[arg] = _has_function(set(), arg)
+      has_func = has_func or FUNCTION_CACHE[arg]
```
Outdent

pgoodman commented on make_abi_library.py:
```diff
@@ -0,0 +1,153 @@
+    if arg not in FUNCTION_CACHE:
+      FUNCTION_CACHE[arg] = _has_function(set(), arg)
+      has_func = has_func or FUNCTION_CACHE[arg]
+      if has_func:
```
Outdent

artemdinaburg commented:
Good catch; outdented.
